### PR TITLE
[RFC] feat(k8s): stop sidecars on otherwise completed Jobs

### PR DESCRIPTION
**Pull request by nojnhuh**
**Description**:

This change adds a new long-lived goroutine to osm-controller to explicitly stop sidecar containers on Pods controlled by Jobs. Currently, Jobs never complete because the sidecars run indefinitely. This is the high-level approach:

- For any `pod-updated` event:
  - If the Pod does not belong to a Job, ignore it
  - If the Pod's status indicates the main containers are still running, ignore it
  - For each running sidecar, send `SIGTERM` to PID 1 with the equivalent of `kubectl exec <pod> -c <container> -- kill 1`
    - Based on how https://hub.docker.com/r/zachomedia/kubernetes-sidecar-terminator works

This handles both `envoy` sidecars and `osm-healthcheck` sidecars (for Pods with `tcpSocket` health probes).

The main restriction this introduces is that any sidecar we intend to inject must include a `kill` command. Envoy's images already include busybox which provides a `kill` command. These changes include changing the base image for `osm-healthcheck` to busybox to give it a `kill` command.

Part of #2316

**Testing done**:

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change?

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)?

### CODE REVIEW

**Comment by steeling** [comment]
What are your thoughts on the wrapping the sidecar proxy (and healthcheck) into a single container. That container could contain a startup script that additionally triggers this, except only watches itself.

This provides a nice platform for us to add more logic into sidecars, hides the healthcheck container, and has a more self-contained mechanism for killing a container.

It scares me a bit that a bug in that code could result in unwanted killing of containers.

File: pkg/k8s/util.go [hunk 1]
```diff
@@ -178,3 +186,79 @@ func NamespacedNameFrom(name string) (types.NamespacedName, error) {
 func IsHeadlessService(svc corev1.Service) bool {
 	return len(svc.Spec.ClusterIP) == 0 || svc.Spec.ClusterIP == corev1.ClusterIPNone
 }
+
+func StopJobSidecars(stop <-chan struct{}, broker *messaging.Broker, config *rest.Config, clientset kubernetes.Interface) {
```
Comment (steeling) [inline pkg/k8s/util.go:190]: nit: we should avoid passing client objects as params when we're not doing so for dependency injection.

Instead, we should maintain a higher level client object (k8s.Controller, or something similar), and place this as a method on that type, which has access to the underlying (dependency injected client).

This maintains a cleaner approach for decoupling from API's and leveraging interfaces.

See https://docs.google.com/document/d/1L_u-XlVU8Fog-uFS3Uf0O2nz8IArQy82cd6Npxm038Y/edit#heading=h.cvqx4tbzt6i7

**Comment by nojnhuh** [comment]
Before putting everything into one sidecar, I got this working as a separate sidecar like osm-healthcheck. I've come across two limitations with this approach that would also affect if everything was in a single sidecar container:
- The Pods' RBAC needs to allow listing/watching Pods to tell when the main containers are done
- Since Envoy redirection is applied to this new sidecar, the sidecar needs to be allowed to communicate with the API server, either by enabling egress or adding an outbound IP range exclusion, or there might be a way to make this more automatic by changing the iptables rules for this specific use case.

We could perhaps get around communicating externally at all if we instead set [`shareProcessNamespace`](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#hosts-namespaces) on Pods and inspect the processes directly. The main limitation there is that it would be impossible for the sidecar to know exactly what the "main" processes are it should wait on. Since there's no real coordination of execution among a Pod's containers, it seems possible that the "main" containers' processes could start and finish before the sidecar has a chance to see what's running in the rest of the Pod. Inspecting the container runtime on the node is probably a better version of this approach, but that is probably similar to the proposed changes to upstream Kubernetes mentioned in #2316 and would probably be outside the scope of something like OSM already running on top of Kubernetes.

Overall, I'm getting the sense the approach the changes on this PR show is simpler than these alternatives.

**Comment by steeling** [comment]
> Before putting everything into one sidecar, I got this working as a separate sidecar like osm-healthcheck. I've come across two limitations with this approach that would also affect if everything was in a single sidecar container:
> 
> * The Pods' RBAC needs to allow listing/watching Pods to tell when the main containers are done
> * Since Envoy redirection is applied to this new sidecar, the sidecar needs to be allowed to communicate with the API server, either by enabling egress or adding an outbound IP range exclusion, or there might be a way to make this more automatic by changing the iptables rules for this specific use case.
> 
> We could perhaps get around communicating externally at all if we instead set [`shareProcessNamespace`](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#hosts-namespaces) on Pods and inspect the processes directly. The main limitation there is that it would be impossible for the sidecar to know exactly what the "main" processes are it should wait on. Since there's no real coordination of execution among a Pod's containers, it seems possible that the "main" containers' processes could start and finish before the sidecar has a chance to see what's running in the rest of the Pod. Inspecting the container runtime on the node is probably a better version of this approach, but that is probably similar to the proposed changes to upstream Kubernetes mentioned in #2316 and would probably be outside the scope of something like OSM already running on top of Kubernetes.
> 
> Overall, I'm getting the sense the approach the changes on this PR show is simpler than these alternatives.

Ya I agree with your conclusions on leveraging the control plane then.

I wonder if it would be even easier to provide thorough documentation for how users can fix this themselves. Something that just hits the proxy's `quitquitquit` page when their app is done. It is a common enough problem though.

We really just need something to tie us over until sidecars get more love from k8s (could be another couple years though)

**Comment by keithmattix** [comment]
So I implemented the quitquitquit solution with a Linux trap when dealing with a different service mesh. It worked pretty well as long as the main container had curl available. It even had retries and timeouts since curl supports it natively

**Comment by nojnhuh** [comment]
> I wonder if it would be even easier to provide thorough documentation for how users can fix this themselves. Something that just hits the proxy's `quitquitquit` page when their app is done. It is a common enough problem though.

> So I implemented the quitquitquit solution with a Linux trap when dealing with a different service mesh. It worked pretty well as long as the main container had curl available. It even had retries and timeouts since curl supports it natively

Trying out `trap`, it's really simple to set up when `sh` is running everything else. I imagine implementing the equivalent functionality should be almost as easy no matter how the container is built and run. Even without curl and using something like `nc` (which is in busybox) you can hand craft the 2-line HTTP request.

The `osm-healthcheck` container doesn't have an HTTP endpoint to stop it, but that would be easy enough to add so it could be stopped similarly. We'd likely want to move that to a different port from the `/osm-healthcheck` endpoint which kubelet interacts with since that has a hole punched for it in iptables, so then the new quit endpoint would only be accessible from `localhost`.

I think documenting this process for now is a good first step to unblock the use case, so I'll close this PR and follow up in #2316.

**Pull request closed without merging**